src/hotspot/share/opto/parseHelper.cpp

```diff
@@ -430,7 +431,12 @@ void VirtualState::print_on(outputStream* os) const {
 }
 #endif
 
-void PEAState::add_new_allocation(Node* obj) {
+void PEAState::add_new_allocation(Node* obj, Compile *c) {
+  const TypeKlassPtr* klass_type = obj->as_Type()->type()->isa_klassptr();
```

Comment:
At this point, `obj` should be CheckCastPPNode and its type is TypeInstPtr. 

In the future, we will extend add_new_allocation() for array object. Therefore, we need to assume its type is TypeOopPtr.  

I don't understand this code snippet. I understand you want to  calculate TypeKlassPtr* from obj. TypeKlasPtr is a subclass of TypePtr but not TypeOopPtr. You may take a look at the class declaration in type.hpp. It also manifests the class hierarchy.

If you want to get TypeKlassPtr*, you would get it from KlassNode. it's alloc->in(KlassNode). There's a API AllocateNode::Ideal_klass() for this purpose. 

From my reading, what you really need is `ciInstanceKlass*`
I have already get ik from TypeOopPtr below. Could you move the gating logic there?  let's focus on TypeInstance for the time being.

Reply:
Yeah, I simplified it a lot. Now do the checks together with where we opt out of throwable class.